### PR TITLE
[t1822] fixed module/plugin loading

### PR DESCRIPTION
**Pull request by secretrobotron**
1. Removed erroneous onModuleReady call in src/editor/module
2. Cleaned up modules loader a tiny bit
3. Removed unnecessary code from src/plugin/*
4. Wrote simple tests for plugin module

### CODE REVIEW

File: src/plugin/module.js [hunk 1]
```diff
@@ -2,229 +2,218 @@
  * If a copy of the MIT license was not distributed with this file, you can
  * obtain one at http://www.mozillapopcorn.org/butter-license.txt */
 
-(function() {
```
Comment (secretrobotron) [inline src/plugin/module.js:5]: Note: there is a lot of strange/duplicated removed/added code in this file because I fixed its indentation and styling.

**Comment by ghost** [comment]
#### From: Bot.io (butyr.org)

---

# Received

Command **cmd_check** from @mjschranz received. Current queue size: 0

Live output at: http://botio.butyr.org:8000/a02a2254aa2bdc5/output.txt

**Comment by ghost** [comment]
#### From: Bot.io (butyr.org)

---

# Success

Full output at http://botio.butyr.org:8000/a02a2254aa2bdc5/output.txt

Total script time: 0.32 mins

**Comment by ghost** [comment]
#### From: Bot.io (butyr.org)

---

# Received

Command **cmd_preview** from @mjschranz received. Current queue size: 0

Live output at: http://botio.butyr.org:8000/c6196a4d2e5cdc4/output.txt

**Comment by ghost** [comment]
#### From: Bot.io (butyr.org)

---

# Success

Full output at http://botio.butyr.org:8000/c6196a4d2e5cdc4/output.txt

Total script time: 0.83 mins

Server is running on http://pr656.butyr.org/templates/test.html

File: test/plugins/index.html [hunk 2]
```diff
@@ -0,0 +1,25 @@
+<!DOCTYPE html>
+<html>
+  <head>
+    <title>Butter Test Suite [Popcorn Wrapper]</title>
```
Comment (mjschranz) [inline test/plugins/index.html]: Copy pasta gone wrong!

File: test/plugins/index.html [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+    <link rel="stylesheet" href="../qunit/qunit.css" type="text/css" media="screen">
+    <script type="text/javascript" src="../qunit/qunit.js"></script>
+    <script type="text/javascript" src="../inject.js"></script>
+    <script type="text/javascript" src="../../src/butter.js"></script>
+    <script type="text/javascript" src="../../external/popcorn-js/popcorn.js"></script>
+    <script type="text/javascript" src="../../external/popcorn-js/modules/player/popcorn.player.js"></script>
+    <script type="text/javascript" src="../../external/popcorn-js/plugins/text/popcorn.text.js"></script>
```
Comment (mjschranz) [inline test/plugins/index.html]: You don't really need the above three scripts included do you?

File: src/ui/ui.js [hunk 4]
```diff
@@ -146,6 +124,28 @@ define( [ "core/eventmanager", "./toggler", "./logo-spinner", "./context-button"
       });
     }
 
+    this.loadIcons = function( icons, resourcesDir ){
+      var icon, img, div;
+
+      for( icon in icons ){
```
Comment (mjschranz) [inline src/ui/ui.js]: space before opening brace

space after <code>for</code>

File: src/ui/ui.js [hunk 5]
```diff
@@ -146,6 +124,28 @@ define( [ "core/eventmanager", "./toggler", "./logo-spinner", "./context-button"
...
+        if( icons.hasOwnProperty( icon ) ){
```
Comment (mjschranz) [inline src/ui/ui.js]: Same as above.
space after <code>if</code>

File: src/plugin/module.js [hunk 6]
```diff
@@ -2,229 +2,218 @@
...
+      for ( var i=0, l=_plugins.length; i<l; ++i ) {
+        if ( _plugins[ i ].name === name ) {
+          return _plugins[ i ];
+        }
+      }
+    };
+
+    DragNDrop.droppable( _container, {
+      drop: function( element ){
+        if( element.getAttribute( "data-butter-draggable-type" ) === "plugin" ){
```
Comment (mjschranz) [inline src/plugin/module.js]: space after <code>if</code> and before opening brace

File: src/plugin/module.js [hunk 7]
```diff
@@ -2,229 +2,218 @@
...
+          var pluginType = element.getAttribute( "data-popcorn-plugin-type" ),
+              plugin = _this.get( pluginType );
+          if( plugin ){
```
Comment (mjschranz) [inline src/plugin/module.js]: same as above

File: src/plugin/module.js [hunk 8]
```diff
@@ -2,229 +2,218 @@
...
+            for( var i=0; i<_listContainer.childNodes.length; ++i ){
```
Comment (mjschranz) [inline src/plugin/module.js]: same as above

File: src/plugin/module.js [hunk 9]
```diff
@@ -2,229 +2,218 @@
...
+              if( _listContainer.childNodes[ i ].getAttribute( "data-popcorn-plugin-type" ) === pluginType ){
```
Comment (mjschranz) [inline src/plugin/module.js]: same as above

File: src/plugin/module.js [hunk 10]
```diff
@@ -2,229 +2,218 @@
...
-      this.clear = function () {
-        while ( _plugins.length > 0 ) {
-          var plugin = _plugins.pop();
+          _plugins.splice( i, 1 );
+          l--;
           _listContainer.removeChild( plugin.element );
+
+          var head = document.getElementsByTagName( "HEAD" )[ 0 ];
+          for ( i = 0, l = head.children.length; i < l; i++ ) {
+            if( head.children[ i ].getAttribute( "src" ) === plugin.path ) {
```
Comment (mjschranz) [inline src/plugin/module.js]: space after <code>if</code>

File: src/plugin/module.js [hunk 11]
```diff
@@ -2,229 +2,218 @@
...
+      }
+
+      var i, l;
+
+      for ( i = 0, l = _plugins.length; i < l; i++ ) {
+        if( _plugins[ i ].name === plugin.name ) {
+          var tracks = butter.tracks;
+          for ( i = 0, l = tracks.length; i < l; i++ ) {
+            var trackEvents = tracks[ i ].trackEvents;
+            for( var k = 0, ln = trackEvents.length - 1; ln >= k; ln-- ) {
```
Comment (mjschranz) [inline src/plugin/module.js]: space after <code>for</code>

File: src/plugin/module.js [hunk 12]
```diff
@@ -2,229 +2,218 @@
...
-            var head = document.getElementsByTagName( "HEAD" )[ 0 ];
-            for ( i = 0, l = head.children.length; i < l; i++ ) {
-              if( head.children[ i ].getAttribute( "src" ) === plugin.path ) {
-                head.removeChild( head.children[ i ] );
+      }
+
+      var i, l;
+
+      for ( i = 0, l = _plugins.length; i < l; i++ ) {
+        if( _plugins[ i ].name === plugin.name ) {
```
Comment (mjschranz) [inline src/plugin/module.js]: space after <code>if</code>

File: src/plugin/module.js [hunk 13]
```diff
@@ -2,229 +2,218 @@
...
 
-      this.remove = function( plugin ) {
+    this.remove = function( plugin ) {
 
-        if( typeof plugin === "string" ) {
-          plugin = this.get( plugin );
-          if( !plugin ) {
-            return;
-          }
+      if( typeof plugin === "string" ) {
```
Comment (mjschranz) [inline src/plugin/module.js]: space after <code>if</code>

File: src/plugin/module.js [hunk 14]
```diff
@@ -2,229 +2,218 @@
...
+          url: plugin.path,
+          check: generatePluginTypeCheckFunction( plugin.type )
+        });
+        newPlugins.push( plugin );
+      }
+
+      butter.loader.load( pluginLoadDescriptors, function(){
+        for( i = 0, l = newPlugins.length; i < l; i++ ) {
+          plugin = newPlugins[ i ];
           if( moduleOptions.defaults && moduleOptions.defaults.indexOf( plugin.type ) > -1 ){
```
Comment (mjschranz) [inline src/plugin/module.js]: You didn't technically add this code, but I'd be happy if you fixed the spacing here too.

File: src/plugin/module.js [hunk 15]
```diff
@@ -2,229 +2,218 @@
...
+          pluginLoadDescriptors = [],
+          plugin,
+          i,
+          l;
+
+      if( ! ( plugins instanceof Array ) ) {
+        plugins = [ plugins ];
+      }
+
+      for( i = 0, l = plugins.length; i < l; i++ ) {
```
Comment (mjschranz) [inline src/plugin/module.js]: space after <code>for</code>

File: src/plugin/module.js [hunk 16]
```diff
@@ -2,229 +2,218 @@
...
+    title.innerHTML = "<span>My Events</span>";
+    _container.appendChild( title );
+    _listWrapper.appendChild( _listContainer );
+    _container.appendChild( _listWrapper );
+
+    var _scrollbar = new Scrollbars.Vertical( _listWrapper, _listContainer );
+    _container.appendChild( _scrollbar.element );
+
+    this._start = function( onModuleReady ){
+      if( butter.ui ){
```
Comment (mjschranz) [inline src/plugin/module.js]: space after <code>if</code> and before opening brace

File: src/plugin/module.js [hunk 17]
```diff
@@ -2,229 +2,218 @@
...
+    saturation = ( ( srcString.charCodeAt( 0 ) + srcString.charCodeAt( 2 ) + srcString.charCodeAt( 4 ) + srcString.charCodeAt( 6 ) ) % 100 ) * 0.05 + 95;
+    lightness = ( ( srcString.charCodeAt( 1 ) + srcString.charCodeAt( 3 ) + srcString.charCodeAt( 5 ) + srcString.charCodeAt( 7 ) ) % 100 ) * 0.20 + 40;
+
+    // bump up reds because they're hard to see
+    if( hue < 20 || hue > 340 ){
+      lightness += 10;
+    }
+
+    // dial back blue/greens a bit
+    if( hue > 160 && hue < 200 ){
```
Comment (mjschranz) [inline src/plugin/module.js]: space after <code>if</code> and before opening brace

File: src/plugin/module.js [hunk 18]
```diff
@@ -2,229 +2,218 @@
...
+    // very simple hashing function
+    while( srcString.length < 9 ){
+      srcString += type;
+    }
+    hue = ( srcString.charCodeAt( 0 ) + srcString.charCodeAt( 3 ) + srcString.charCodeAt( 6 ) ) % ( ( srcString.charCodeAt( 8) * 5 ) % 360 );
+    saturation = ( ( srcString.charCodeAt( 0 ) + srcString.charCodeAt( 2 ) + srcString.charCodeAt( 4 ) + srcString.charCodeAt( 6 ) ) % 100 ) * 0.05 + 95;
+    lightness = ( ( srcString.charCodeAt( 1 ) + srcString.charCodeAt( 3 ) + srcString.charCodeAt( 5 ) + srcString.charCodeAt( 7 ) ) % 100 ) * 0.20 + 40;
+
+    // bump up reds because they're hard to see
+    if( hue < 20 || hue > 340 ){
```
Comment (mjschranz) [inline src/plugin/module.js]: space after <code>if</code> and before opening brace

File: src/plugin/module.js [hunk 19]
```diff
@@ -2,229 +2,218 @@
...
+
+  __newStyleSheet.type = "text/css";
+  __newStyleSheet.media = "screen";
+  __newStyleSheet.setAttribute( "data-butter-exclude", "true" );
+
+  function colourHashFromType( type ){
+    var hue = 0, saturation = 0, lightness = 0, srcString = type;
+
+    // very simple hashing function
+    while( srcString.length < 9 ){
```
Comment (mjschranz) [inline src/plugin/module.js]: space after <code>while</code> and before opening brace

File: src/plugin/module.js [hunk 20]
```diff
@@ -2,229 +2,218 @@
...
+    var _scrollbar = new Scrollbars.Vertical( _listWrapper, _listContainer );
+    _container.appendChild( _scrollbar.element );
+
+    this._start = function( onModuleReady ){
+      if( butter.ui ){
+        document.head.appendChild( __newStyleSheet );
+        butter.ui.areas.tools.addComponent( _container );
+        PluginList( butter );
+      }
+      if( moduleOptions && moduleOptions.plugins ){
```
Comment (mjschranz) [inline src/plugin/module.js]: same as above

File: src/plugin/module.js [hunk 21]
```diff
@@ -2,229 +2,218 @@
...
+    title.className = "title";
+    title.innerHTML = "<span>My Events</span>";
+    _container.appendChild( title );
+    _listWrapper.appendChild( _listContainer );
+    _container.appendChild( _listWrapper );
+
+    var _scrollbar = new Scrollbars.Vertical( _listWrapper, _listContainer );
+    _container.appendChild( _scrollbar.element );
+
+    this._start = function( onModuleReady ){
```
Comment (mjschranz) [inline src/plugin/module.js]: space before opening brace

File: src/plugin/module.js [hunk 22]
```diff
@@ -2,229 +2,218 @@
...
+
+      var i, l;
+
+      for ( i = 0, l = _plugins.length; i < l; i++ ) {
+        if( _plugins[ i ].name === plugin.name ) {
+          var tracks = butter.tracks;
+          for ( i = 0, l = tracks.length; i < l; i++ ) {
+            var trackEvents = tracks[ i ].trackEvents;
+            for( var k = 0, ln = trackEvents.length - 1; ln >= k; ln-- ) {
+              if( trackEvents[ ln ].type === plugin.name ) {
```
Comment (mjschranz) [inline src/plugin/module.js]: space after <code>if</code>

File: src/plugin/module.js [hunk 23]
```diff
@@ -2,229 +2,218 @@
...
-              for( var i=0; i<_listContainer.childNodes.length; ++i ){
-                if( _listContainer.childNodes[ i ].getAttribute( "data-popcorn-plugin-type" ) === pluginType ){
-                  return;
-                }
+      }
+
+      _scrollbar.update();
+    };
+
+    this.clear = function () {
```
Comment (mjschranz) [inline src/plugin/module.js]: why the space after <code>function</code> here?

File: src/ui/ui.js [hunk 24]
```diff
@@ -146,6 +124,28 @@ define( [ "core/eventmanager", "./toggler", "./logo-spinner", "./context-button"
       });
     }
 
+    this.loadIcons = function( icons, resourcesDir ){
```
Comment (mjschranz) [inline src/ui/ui.js]: space before <code>{</code>

File: test/plugins/plugins.unit.js [hunk 25]
```diff
@@ -0,0 +1,69 @@
+/*global Butter,asyncTest,equal,start,ok*/
+require( [], function() {
```
Comment (mjschranz) [inline test/plugins/plugins.unit.js]: Why have require here if you aren't loading anything?

I'm really curious. Trying to figure out if similar issues in the other tests I wrote are what's causing some problems on test swarm

File: test/plugins/plugins.unit.js [hunk 26]
```diff
@@ -0,0 +1,69 @@
...
+
+  // All modules that create Butter objects (e.g., Butter())
+  // should use this lifecycle, and call rememberButter() for all
+  // created butter instances.  Any created using createButter()
+  // already have it done automatically.
+  var butterLifeCycle = (function(){
```
Comment (mjschranz) [inline test/plugins/plugins.unit.js:7]: Again, another general curiosity. Is this something needed for such a small amount of tests? If so, I need to get this in for mine as well.

**Comment by ghost** [comment]
#### From: Bot.io (butyr.org)

---

# Received

Command **cmd_try** from @mjschranz received. Current queue size: 0

Live output at: http://botio.butyr.org:8000/25038ec4845dce6/output.txt

**Comment by ghost** [comment]
#### From: Bot.io (butyr.org)

---

# Success

Full output at http://botio.butyr.org:8000/25038ec4845dce6/output.txt

Total script time: 1.33 mins

Your job has been submitted to the swarm! see http://norway.proximity.on.ca/user/butter

File: test/plugins/plugins.unit.js (hunk 25 above)
Comment (secretrobotron) [inline test/plugins/plugins.unit.js]: They might be. I wasn't sure what you were doing here, but if they're not required (no pun intended), I'll remove them here.

File: test/plugins/plugins.unit.js (hunk 26 above)
Comment (secretrobotron) [inline test/plugins/plugins.unit.js:7]: I'm just preparing for having more tests, really. Seems like great boilerplate code to have in all test modules.

**Comment by ghost** [comment]
#### From: Bot.io (butyr.org)

---

# Received

Command **cmd_cleanup** from @mjschranz received. Current queue size: 0

Live output at: http://botio.butyr.org:8000/d161b851bff7b72/output.txt

**Comment by ghost** [comment]
#### From: Bot.io (butyr.org)

---

# Success

Full output at http://botio.butyr.org:8000/d161b851bff7b72/output.txt

Total script time: 0.22 mins

**Comment by ghost** [comment]
#### From: Bot.io (butyr.org)

---

# Received

Command **cmd_preview** from @mjschranz received. Current queue size: 0

Live output at: http://botio.butyr.org:8000/88b4c4b70945401/output.txt

**Comment by ghost** [comment]
#### From: Bot.io (butyr.org)

---

# Success

Full output at http://botio.butyr.org:8000/88b4c4b70945401/output.txt

Total script time: 0.86 mins

Server is running on http://pr656.butyr.org/templates/test.html

File: src/plugin/module.js [hunk 27]
```diff
@@ -2,229 +2,218 @@
...
-                head.removeChild( head.children[ i ] );
+      }
+
+      var i, l;
+
+      for ( i = 0, l = _plugins.length; i < l; i++ ) {
+        if ( _plugins[ i ].name === plugin.name ) {
+          var tracks = butter.tracks;
+          for ( i = 0, l = tracks.length; i < l; i++ ) {
+            var trackEvents = tracks[ i ].trackEvents;
```
Comment (dseif) [inline src/plugin/module.js]: Can you declare trackEvents and tracks on line 148 instead of inside the for loop

**Comment by ghost** [comment]
#### From: Bot.io (butyr.org)

---

# Received

Command **cmd_check** from @dseif received. Current queue size: 0

Live output at: http://botio.butyr.org:8000/0f275b251ea92de/output.txt

**Comment by ghost** [comment]
#### From: Bot.io (butyr.org)

---

# Success

Full output at http://botio.butyr.org:8000/0f275b251ea92de/output.txt

Total script time: 0.31 mins

**Comment by ghost** [comment]
#### From: Bot.io (butyr.org)

---

# Received

Command **cmd_preview** from @dseif received. Current queue size: 0

Live output at: http://botio.butyr.org:8000/afbb5f0454cd907/output.txt

**Comment by ghost** [comment]
#### From: Bot.io (butyr.org)

---

# Success

Full output at http://botio.butyr.org:8000/afbb5f0454cd907/output.txt

Total script time: 0.90 mins

Server is running on http://pr656.butyr.org/templates/test.html

**Pull request closed without merging**